Assets/Mirror/Runtime/NetworkWriter.cs

```diff
@@ -19,28 +20,74 @@ public class NetworkWriter
 
         // 'int' is the best type for .Position. 'short' is too small if we send >32kb which would result in negative .Position
         // -> converting long to int is fine until 2GB of data (MAX_INT), so we don't have to worry about overflows here
-        public int Position;
-
+        int position;
         int length;
 
-        public int Length
+        public int Length => length;
+
+        public int Position
         {
-            get => length;
-            private set
+            get => position;
+            set
             {
-                EnsureCapacity(value);
-                length = value;
-                if (Position > length)
-                    Position = length;
+                position = value;
+                EnsureLength(value);
+            }
+        }
+
+        /// <summary>
+        /// Reset both the position and length of the stream
+        /// </summary>
+        /// <remarks>
+        /// Leaves the capacity the same so that we can reuse this writer without extra allocations
+        /// </remarks>
+        public void Reset()
+        {
+            position = 0;
+            length = 0;
+        }
+
+        /// <summary>
+        /// Sets length, moves position if it is greater than new length
+        /// </summary>
+        /// <param name="value"></param>
+        /// <remarks>
+        /// Zeros out any extra length created by setlength
+        /// </remarks>
+        public void SetLength(int value)
+        {
+            int oldLength = length;
+            
+            // use EnsureCapacity 
+            EnsureLength(value);
+
+            // zero out new length
+            if (oldLength < value)
+            {
+                for (int i = oldLength; i < value; i++)
+                {
+                    buffer[i] = 0;
+                }
```

Comment:
I believe Array.Clear is optimized with SIMD or other CPU secret sauce
```suggestion
                Array.Clear(buffer, oldLength, value - oldLength);
```